File: frontend/src/lib/api/social.ts
import { apiFetch } from "./client";

export type PlanRole = "viewer" | "editor" | "owner";
export type InvitableRole = Exclude<PlanRole, "owner">;
export type FriendshipStatus = "pending" | "accepted" | "rejected";

export interface ProfileSummary {
  id: string;
  username: string;
  display_name: string | null;
  avatar_url: string | null;
}

export interface Friendship {
  id: string;
  requester_id: string;
  addressee_id: string;
  status: FriendshipStatus;
  created_at: string;
  other: ProfileSummary;
}

export interface PlanMember {
  plan_id: string;
  user_id: string;
  role: PlanRole;
  joined_at: string;
  profile: ProfileSummary;
}

export interface PlanInvite {
  id: string;
  plan_id: string;
  token: string;
  role: PlanRole;
  expires_at: string | null;
  max_uses: number | null;
  uses: number;
  created_by: string | null;
  created_at: string;
}

export interface CreateInvitePayload {
  role: InvitableRole;
  expires_in_hours?: number | null;
  max_uses?: number | null;
}

export interface InviteAcceptResponse {
  plan_id: string;
  role: PlanRole;
}

export const searchUsers = (q: string): Promise<ProfileSummary[]> => {
  const params = new URLSearchParams({ q });
  return apiFetch<ProfileSummary[]>(`/social/users/search?${params.toString()}`);
};

export const listFriends = (): Promise<Friendship[]> =>
  apiFetch<Friendship[]>("/social/friends");

export const listIncomingRequests = (): Promise<Friendship[]> =>
  apiFetch<Friendship[]>("/social/friends/incoming");

export const listOutgoingRequests = (): Promise<Friendship[]> =>
  apiFetch<Friendship[]>("/social/friends/outgoing");

export const sendFriendRequest = (username: string): Promise<Friendship> =>
  apiFetch<Friendship>("/social/friends/request", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username }),
  });

export const acceptFriendRequest = (requestId: string): Promise<Friendship> =>
  apiFetch<Friendship>(`/social/friends/accept/${requestId}`, { method: "POST" });

export const rejectFriendRequest = (requestId: string): Promise<Friendship> =>
  apiFetch<Friendship>(`/social/friends/reject/${requestId}`, { method: "POST" });

export const cancelFriendRequest = (requestId: string): Promise<void> =>
  apiFetch<void>(`/social/friends/requests/${requestId}`, { method: "DELETE" });

export const removeFriend = (userId: string): Promise<void> =>
  apiFetch<void>(`/social/friends/${userId}`, { method: "DELETE" });

export const listPlanMembers = (planId: string): Promise<PlanMember[]> =>
  apiFetch<PlanMember[]>(`/plans/${planId}/members`);

export const addPlanMember = (
  planId: string,
  userId: string,
  role: InvitableRole,
): Promise<PlanMember> =>
  apiFetch<PlanMember>(`/plans/${planId}/members`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ user_id: userId, role }),
  });

export const getMyPlanRole = (
  planId: string,
  accessToken?: string | null,
): Promise<{ role: PlanRole }> =>
  apiFetch<{ role: PlanRole }>(`/plans/${planId}/role`, undefined, accessToken);

export const updatePlanMemberRole = (
  planId: string,
  userId: string,
  role: PlanRole,
): Promise<PlanMember> =>
  apiFetch<PlanMember>(`/plans/${planId}/members/${userId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ role }),
  });

export const removePlanMember = (planId: string, userId: string): Promise<void> =>
  apiFetch<void>(`/plans/${planId}/members/${userId}`, { method: "DELETE" });

export const listPlanInvites = (planId: string): Promise<PlanInvite[]> =>
  apiFetch<PlanInvite[]>(`/plans/${planId}/invites`);

export const createPlanInvite = (
  planId: string,
  payload: CreateInvitePayload,
): Promise<PlanInvite> =>
  apiFetch<PlanInvite>(`/plans/${planId}/invites`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

export const revokePlanInvite = (planId: string, inviteId: string): Promise<void> =>
  apiFetch<void>(`/plans/${planId}/invites/${inviteId}`, { method: "DELETE" });

export const acceptInvite = (
  token: string,
  accessToken?: string | null,
): Promise<InviteAcceptResponse> =>
  apiFetch<InviteAcceptResponse>(
    `/invite/${token}/accept`,
    { method: "POST" },
    accessToken,
  );

File: backend/app/schemas/social.py
"""Pydantic models for social endpoints — friends, plan members, invites,
comments, reactions, ratings, activity."""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

PlanRole = Literal["viewer", "editor", "owner"]
InvitableRole = Literal["viewer", "editor"]
FriendshipStatus = Literal["pending", "accepted", "rejected"]
ReactionKind = Literal["like", "dislike", "love", "bookmark"]


class ProfileSummary(BaseModel):
    id: str
    username: str
    display_name: str | None = None
    avatar_url: str | None = None


class FriendshipResponse(BaseModel):
    id: str
    requester_id: str
    addressee_id: str
    status: FriendshipStatus
    created_at: str
    other: ProfileSummary


class FriendRequestCreate(BaseModel):
    username: str = Field(..., min_length=1)


class PlanMemberResponse(BaseModel):
    plan_id: str
    user_id: str
    role: PlanRole
    joined_at: str
    profile: ProfileSummary


class PlanMemberUpdate(BaseModel):
    role: PlanRole


class PlanMemberCreate(BaseModel):
    user_id: str
    role: InvitableRole = "viewer"


class PlanInviteResponse(BaseModel):
    id: str
    plan_id: str
    token: str
    role: PlanRole
    expires_at: str | None = None
    max_uses: int | None = None
    uses: int
    created_by: str | None = None
    created_at: str


class PlanInviteCreate(BaseModel):
    role: InvitableRole = "viewer"
    expires_in_hours: int | None = Field(None, ge=1, le=24 * 365)
    max_uses: int | None = Field(None, ge=1)


class InviteAcceptResponse(BaseModel):
    plan_id: str
    role: PlanRole


class CommentResponse(BaseModel):
    id: str
    plan_id: str
    plan_item_id: str | None = None
    parent_id: str | None = None
    author_id: str | None = None
    body: str
    created_at: str
    updated_at: str
    deleted_at: str | None = None
    author: ProfileSummary | None = None


class CommentCreate(BaseModel):
    body: str = Field(..., min_length=1, max_length=4000)
    plan_item_id: str | None = None
    parent_id: str | None = None


class CommentUpdate(BaseModel):
    body: str = Field(..., min_length=1, max_length=4000)


class ReactionResponse(BaseModel):
    plan_item_id: str
    user_id: str
    kind: ReactionKind
    created_at: str


class ReactionCreate(BaseModel):
    kind: ReactionKind


class RatingResponse(BaseModel):
    plan_item_id: str
    user_id: str
    stars: int = Field(..., ge=1, le=5)
    created_at: str
    updated_at: str


class RatingUpsert(BaseModel):
    stars: int = Field(..., ge=1, le=5)


class ActivityResponse(BaseModel):
    id: str
    plan_id: str
    actor_id: str | None = None
    kind: str
    payload: dict[str, Any] | None = None
    created_at: str
    actor: ProfileSummary | None = None
